Upload fails often in 2.7

Report from user on googlegroups
He says "Very often: Upload failed"

And has logs attached. Unfortunately I can't share them here without permission, but anyone who has signed the NDA that allows them to access our private group is welcome to check it there.

There are two issues here:

1. The logs don't provide any meaningful information. It is completely cluttered by system logs, etc. This has been the case for logs provided by everyone else as well. Is it at all possible for us to filter for our app's logs programmatically in the "send logs" feature?

2. The upload failing very often. I plan to ask him where he is uploading from (sd card, etc), and for screenshots if he can provide them. What other information do you guys think would be useful to solve the problem? @maskaravivek @psh @nicolas-raoul @neslihanturan 

He isn't the only one with it - a recent Play Store review was from a user who encountered the same problem. So I would probably assign this issue a fairly high priority since this is our core functionality.

I had a failure happen to me while testing the 2.8-release branch on prodDebug. Associated logs:

```
07-17 18:41:40.103 6243-6299/fr.free.nrw.commons.debug D/UploadService$Notificat: Uploaded 1713889 of 2516894
07-17 18:41:40.117 6243-6299/fr.free.nrw.commons.debug D/UploadService$Notificat: Uploaded 1791713 of 2516894
07-17 18:41:40.131 6243-6299/fr.free.nrw.commons.debug D/UploadService$Notificat: Uploaded 1869537 of 2516894
07-17 18:41:41.063 6243-6299/fr.free.nrw.commons.debug D/UploadService$Notificat: Uploaded 1947361 of 2516894
07-17 18:41:42.023 6243-6299/fr.free.nrw.commons.debug D/UploadService$Notificat: Uploaded 2025185 of 2516894
07-17 18:41:42.039 6243-6299/fr.free.nrw.commons.debug D/UploadService$Notificat: Uploaded 2103009 of 2516894
07-17 18:41:43.463 6243-6299/fr.free.nrw.commons.debug D/UploadService$Notificat: Uploaded 2180833 of 2516894
07-17 18:41:43.479 6243-6299/fr.free.nrw.commons.debug D/UploadService$Notificat: Uploaded 2258657 of 2516894
07-17 18:41:44.423 6243-6299/fr.free.nrw.commons.debug D/UploadService$Notificat: Uploaded 2336481 of 2516894
07-17 18:41:44.761 6243-6290/fr.free.nrw.commons.debug D/EGL_emulation: eglMakeCurrent: 0xb2785600: ver 2 0 (tinfo 0xb2783470)
07-17 18:41:44.772 6243-6290/fr.free.nrw.commons.debug D/EGL_emulation: eglMakeCurrent: 0xb2785600: ver 2 0 (tinfo 0xb2783470)
07-17 18:41:44.779 6243-6290/fr.free.nrw.commons.debug D/EGL_emulation: eglMakeCurrent: 0xb2785600: ver 2 0 (tinfo 0xb2783470)
07-17 18:41:44.789 6243-6290/fr.free.nrw.commons.debug D/EGL_emulation: eglMakeCurrent: 0xb2785600: ver 2 0 (tinfo 0xb2783470)
07-17 18:41:44.795 6243-6290/fr.free.nrw.commons.debug D/EGL_emulation: eglMakeCurrent: 0xb2785600: ver 2 0 (tinfo 0xb2783470)
07-17 18:41:44.806 6243-6290/fr.free.nrw.commons.debug D/OpenGLRenderer: endAllActiveAnimators on 0x96b09200 (RippleDrawable) with handle 0x96bc1780
07-17 18:41:45.083 1661-2349/system_process I/WindowManager: Destroying surface Surface(name=fr.free.nrw.commons.debug/fr.free.nrw.commons.upload.ShareActivity) called by com.android.server.wm.WindowStateAnimator.destroySurface:2016 com.android.server.wm.WindowStateAnimator.destroySurfaceLocked:882 com.android.server.wm.WindowState.removeLocked:1456 com.android.server.wm.WindowManagerService.removeWindowInnerLocked:2484 com.android.server.wm.WindowManagerService.removeWindowLocked:2442 com.android.server.wm.WindowManagerService.removeWindowLocked:2311 com.android.server.wm.WindowManagerService.removeWindow:2306 com.android.server.wm.Session.remove:202 
07-17 18:41:45.649 6243-6474/fr.free.nrw.commons.debug D/ModificationsSyncAdapte: Found 1 modifications to execute
07-17 18:41:46.348 6243-6299/fr.free.nrw.commons.debug D/UploadService$Notificat: Uploaded 2414305 of 2516894
07-17 18:41:46.363 6243-6299/fr.free.nrw.commons.debug D/UploadService$Notificat: Uploaded 2492129 of 2516894
07-17 18:41:46.379 6243-6299/fr.free.nrw.commons.debug D/UploadService$Notificat: Uploaded 2516894 of 2516894
07-17 18:41:49.707 6243-6243/fr.free.nrw.commons.debug W/View: requestLayout() improperly called by android.widget.TextView{9327d20 V.ED..... ......ID 12,36-203,73 #7f0f013e app:id/contributionTitle} during layout: running second layout pass
    requestLayout() improperly called by android.widget.TextView{35f36d9 V.ED..... ......ID 395,171-477,363 #7f0f013a app:id/contributionSequenceNumber} during layout: running second layout pass
07-17 18:41:56.689 6243-6299/fr.free.nrw.commons.debug E/WTF: Result: org.mediawiki.api.ApiResult@61fe214
07-17 18:41:56.694 6243-6299/fr.free.nrw.commons.debug E/ApacheHttpClientMediaWi: mustbeloggedin
07-17 18:41:56.694 6243-6299/fr.free.nrw.commons.debug D/UploadService: Response is fr.free.nrw.commons.mwapi.UploadResult@1ed8abd
    Contribution upload failed. Wikidata entity won't be edited
07-17 18:41:57.289 6243-6480/fr.free.nrw.commons.debug D/ModificationsSyncAdapte: Found 1 modifications to execute
07-17 18:41:57.706 6243-6243/fr.free.nrw.commons.debug W/View: requestLayout() improperly called by android.widget.TextView{56d15b9 V.ED..... ......ID 12,36-203,73 #7f0f013e app:id/contributionTitle} during layout: running second layout pass
    requestLayout() improperly called by android.widget.TextView{226cdfe V.ED..... ......ID 395,171-477,363 #7f0f013a app:id/contributionSequenceNumber} during layout: running second layout pass
    requestLayout() improperly called by android.widget.TextView{dbd495f V.ED..... ......ID 12,12-68,41 #7f0f013d app:id/contributionState} during layout: running second layout pass
07-17 18:42:00.623 6243-6243/fr.free.nrw.commons.debug D/MediaDetailFragment: MediaDetailFragment ready to display details
07-17 18:42:00.643 6243-6243/fr.free.nrw.commons.debug D/MediaDetailFragment: MediaDetailFragment ready to display details
07-17 18:42:00.644 6243-6243/fr.free.nrw.commons.debug D/MediaDetailFragment: MediaDetailFragment ready to display details
07-17 18:42:01.113 6243-6452/fr.free.nrw.commons.debug D/MediaDataExtractor: Nominated for deletion: false
07-17 18:42:01.589 6243-6357/fr.free.nrw.commons.debug D/MediaDataExtractor: Nominated for deletion: false
07-17 18:42:01.618 6243-6452/fr.free.nrw.commons.debug D/MediaDetailFragment: java.io.IOException: org.xml.sax.SAXParseException: Unexpected end of document
        at fr.free.nrw.commons.MediaDataExtractor.processWikiParseTree(MediaDataExtractor.java:104)
        at fr.free.nrw.commons.MediaDataExtractor.fetch(MediaDataExtractor.java:77)
        at fr.free.nrw.commons.media.MediaDetailFragment$3.doInBackground(MediaDetailFragment.java:252)
        at fr.free.nrw.commons.media.MediaDetailFragment$3.doInBackground(MediaDetailFragment.java:241)
        at android.os.AsyncTask$2.call(AsyncTask.java:304)
        at java.util.concurrent.FutureTask.run(FutureTask.java:237)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
        at java.lang.Thread.run(Thread.java:761)
     Caused by: org.xml.sax.SAXParseException: Unexpected end of document
        at org.apache.harmony.xml.parsers.DocumentBuilderImpl.parse(DocumentBuilderImpl.java:125)
        at javax.xml.parsers.DocumentBuilder.parse(DocumentBuilder.java:107)
        at fr.free.nrw.commons.MediaDataExtractor.processWikiParseTree(MediaDataExtractor.java:100)
        at fr.free.nrw.commons.MediaDataExtractor.fetch(MediaDataExtractor.java:77) 
        at fr.free.nrw.commons.media.MediaDetailFragment$3.doInBackground(MediaDetailFragment.java:252) 
        at fr.free.nrw.commons.media.MediaDetailFragment$3.doInBackground(MediaDetailFragment.java:241) 
        at android.os.AsyncTask$2.call(AsyncTask.java:304) 
        at java.util.concurrent.FutureTask.run(FutureTask.java:237) 
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607) 
        at java.lang.Thread.run(Thread.java:761) 
07-17 18:42:01.625 6243-6243/fr.free.nrw.commons.debug D/MediaDetailFragment: Failed to load photo details.
07-17 18:42:02.109 6243-6357/fr.free.nrw.commons.debug D/MediaDetailFragment: java.io.IOException: org.xml.sax.SAXParseException: Unexpected end of document
        at fr.free.nrw.commons.MediaDataExtractor.processWikiParseTree(MediaDataExtractor.java:104)
        at fr.free.nrw.commons.MediaDataExtractor.fetch(MediaDataExtractor.java:77)
        at fr.free.nrw.commons.media.MediaDetailFragment$3.doInBackground(MediaDetailFragment.java:252)
        at fr.free.nrw.commons.media.MediaDetailFragment$3.doInBackground(MediaDetailFragment.java:241)
        at android.os.AsyncTask$2.call(AsyncTask.java:304)
        at java.util.concurrent.FutureTask.run(FutureTask.java:237)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
        at java.lang.Thread.run(Thread.java:761)
     Caused by: org.xml.sax.SAXParseException: Unexpected end of document
        at org.apache.harmony.xml.parsers.DocumentBuilderImpl.parse(DocumentBuilderImpl.java:125)
        at javax.xml.parsers.DocumentBuilder.parse(DocumentBuilder.java:107)
        at fr.free.nrw.commons.MediaDataExtractor.processWikiParseTree(MediaDataExtractor.java:100)
        at fr.free.nrw.commons.MediaDataExtractor.fetch(MediaDataExtractor.java:77) 
        at fr.free.nrw.commons.media.MediaDetailFragment$3.doInBackground(MediaDetailFragment.java:252) 
        at fr.free.nrw.commons.media.MediaDetailFragment$3.doInBackground(MediaDetailFragment.java:241) 
        at android.os.AsyncTask$2.call(AsyncTask.java:304) 
        at java.util.concurrent.FutureTask.run(FutureTask.java:237) 
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607) 
        at java.lang.Thread.run(Thread.java:761) 
07-17 18:42:02.115 6243-6452/fr.free.nrw.commons.debug D/MediaDataExtractor: Nominated for deletion: false
07-17 18:42:02.125 6243-6243/fr.free.nrw.commons.debug D/MediaDetailFragment: Failed to load photo details.
07-17 18:42:02.623 6243-6452/fr.free.nrw.commons.debug D/MediaDetailFragment: java.io.IOException: org.xml.sax.SAXParseException: Unexpected end of document
        at fr.free.nrw.commons.MediaDataExtractor.processWikiParseTree(MediaDataExtractor.java:104)
        at fr.free.nrw.commons.MediaDataExtractor.fetch(MediaDataExtractor.java:77)
        at fr.free.nrw.commons.media.MediaDetailFragment$3.doInBackground(MediaDetailFragment.java:252)
        at fr.free.nrw.commons.media.MediaDetailFragment$3.doInBackground(MediaDetailFragment.java:241)
        at android.os.AsyncTask$2.call(AsyncTask.java:304)
        at java.util.concurrent.FutureTask.run(FutureTask.java:237)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
        at java.lang.Thread.run(Thread.java:761)
     Caused by: org.xml.sax.SAXParseException: Unexpected end of document
        at org.apache.harmony.xml.parsers.DocumentBuilderImpl.parse(DocumentBuilderImpl.java:125)
        at javax.xml.parsers.DocumentBuilder.parse(DocumentBuilder.java:107)
        at fr.free.nrw.commons.MediaDataExtractor.processWikiParseTree(MediaDataExtractor.java:100)
        at fr.free.nrw.commons.MediaDataExtractor.fetch(MediaDataExtractor.java:77) 
        at fr.free.nrw.commons.media.MediaDetailFragment$3.doInBackground(MediaDetailFragment.java:252) 
        at fr.free.nrw.commons.media.MediaDetailFragment$3.doInBackground(MediaDetailFragment.java:241) 
        at android.os.AsyncTask$2.call(AsyncTask.java:304) 
        at java.util.concurrent.FutureTask.run(FutureTask.java:237) 
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607) 
        at java.lang.Thread.run(Thread.java:761) 
07-17 18:42:02.624 6243-6243/fr.free.nrw.commons.debug D/MediaDetailFragment: Failed to load photo details.
```

3rd test, via Gallery > selecting a photo stored on external storage (not Google Photos):

```
07-17 18:52:19.148 6243-6299/fr.free.nrw.commons.debug D/UploadService$Notificat: Uploaded 108297 of 118932
07-17 18:52:19.164 6243-6299/fr.free.nrw.commons.debug D/UploadService$Notificat: Uploaded 112393 of 118932
07-17 18:52:19.186 6243-6299/fr.free.nrw.commons.debug D/UploadService$Notificat: Uploaded 116489 of 118932
07-17 18:52:19.206 6243-6299/fr.free.nrw.commons.debug D/UploadService$Notificat: Uploaded 118932 of 118932
07-17 18:52:19.325 6243-6290/fr.free.nrw.commons.debug D/EGL_emulation: eglMakeCurrent: 0xb2785600: ver 2 0 (tinfo 0xb2783470)
07-17 18:52:19.337 6243-6290/fr.free.nrw.commons.debug D/EGL_emulation: eglMakeCurrent: 0xb2785600: ver 2 0 (tinfo 0xb2783470)
07-17 18:52:19.348 6243-6290/fr.free.nrw.commons.debug D/EGL_emulation: eglMakeCurrent: 0xb2785600: ver 2 0 (tinfo 0xb2783470)
07-17 18:52:19.356 6243-6290/fr.free.nrw.commons.debug D/EGL_emulation: eglMakeCurrent: 0xb2785600: ver 2 0 (tinfo 0xb2783470)
07-17 18:52:20.224 6243-6290/fr.free.nrw.commons.debug D/EGL_emulation: eglMakeCurrent: 0xb2785600: ver 2 0 (tinfo 0xb2783470)
07-17 18:52:20.247 6243-6290/fr.free.nrw.commons.debug D/EGL_emulation: eglMakeCurrent: 0xb2785600: ver 2 0 (tinfo 0xb2783470)
07-17 18:52:20.252 1661-1957/system_process I/WindowManager: Destroying surface Surface(name=fr.free.nrw.commons.debug/fr.free.nrw.commons.upload.ShareActivity) called by com.android.server.wm.WindowStateAnimator.destroySurface:2016 com.android.server.wm.WindowStateAnimator.destroySurfaceLocked:882 com.android.server.wm.WindowState.removeLocked:1456 com.android.server.wm.WindowManagerService.removeWindowInnerLocked:2484 com.android.server.wm.WindowManagerService.removeWindowLocked:2442 com.android.server.wm.WindowManagerService.removeWindowLocked:2311 com.android.server.wm.WindowManagerService.removeWindow:2306 com.android.server.wm.Session.remove:202 
07-17 18:52:20.301 6243-6290/fr.free.nrw.commons.debug D/EGL_emulation: eglMakeCurrent: 0xb2785600: ver 2 0 (tinfo 0xb2783470)
07-17 18:52:20.457 6243-6299/fr.free.nrw.commons.debug E/WTF: Result: org.mediawiki.api.ApiResult@53f561f
07-17 18:52:20.460 6243-6299/fr.free.nrw.commons.debug E/ApacheHttpClientMediaWi: mustbeloggedin
07-17 18:52:20.460 6243-6299/fr.free.nrw.commons.debug D/UploadService: Response is fr.free.nrw.commons.mwapi.UploadResult@32f4d6c
    Contribution upload failed. Wikidata entity won't be edited
07-17 18:52:20.862 1661-1692/system_process I/WindowManager: Destroying surface Surface(name=fr.free.nrw.commons.debug/fr.free.nrw.commons.upload.ShareActivity) called by com.android.server.wm.WindowStateAnimator.destroySurface:2016 com.android.server.wm.WindowStateAnimator.destroySurfaceLocked:882 com.android.server.wm.WindowState.removeLocked:1456 com.android.server.wm.WindowManagerService.removeWindowInnerLocked:2484 com.android.server.wm.WindowManagerService.removeWindowLocked:2442 com.android.server.wm.WindowManagerService.removeWindowLocked:2311 com.android.server.wm.WindowManagerService.removeWindow:2306 com.android.server.wm.Session.remove:202 
07-17 18:52:21.037 6243-7343/fr.free.nrw.commons.debug D/ModificationsSyncAdapte: Found 1 modifications to execute
07-17 18:52:21.374 6243-6243/fr.free.nrw.commons.debug W/View: requestLayout() improperly called by android.widget.TextView{b411e96 V.ED..... ......ID 12,41-359,78 #7f0f013e app:id/contributionTitle} during layout: running second layout pass
    requestLayout() improperly called by android.widget.TextView{fea17 V.ED..... ......ID 395,171-477,363 #7f0f013a app:id/contributionSequenceNumber} during layout: running second layout pass
    requestLayout() improperly called by android.widget.TextView{4e3de04 V.ED..... ......ID 12,12-68,41 #7f0f013d app:id/contributionState} during layout: running second layout pass
07-17 18:52:23.390 6243-6243/fr.free.nrw.commons.debug W/View: requestLayout() improperly called by android.widget.TextView{f7cd034 V.ED..... ......ID 12,41-359,78 #7f0f013e app:id/contributionTitle} during layout: running second layout pass
    requestLayout() improperly called by android.widget.TextView{bccae5d V.ED..... ......ID 395,171-477,363 #7f0f013a app:id/contributionSequenceNumber} during layout: running second layout pass
07-17 18:52:23.391 6243-6243/fr.free.nrw.commons.debug W/View: requestLayout() improperly called by android.widget.TextView{c34c3d2 V.ED..... ......ID 12,12-68,41 #7f0f013d app:id/contributionState} during layout: running second layout pass
07-17 18:52:23.840 6243-6243/fr.free.nrw.commons.debug W/View: requestLayout() improperly called by android.widget.TextView{f7cd034 V.ED..... ......ID 12,41-359,78 #7f0f013e app:id/contributionTitle} during layout: running second layout pass
    requestLayout() improperly called by android.widget.TextView{bccae5d V.ED..... ......ID 395,171-477,363 #7f0f013a app:id/contributionSequenceNumber} during layout: running second layout pass
    requestLayout() improperly called by android.widget.TextView{c34c3d2 V.ED..... ......ID 12,12-68,41 #7f0f013d app:id/contributionState} during layout: running second layout pass
07-17 18:52:26.270 6243-6243/fr.free.nrw.commons.debug D/MediaDetailFragment: MediaDetailFragment ready to display details
07-17 18:52:26.288 6243-6243/fr.free.nrw.commons.debug D/MediaDetailFragment: MediaDetailFragment ready to display details
    MediaDetailFragment ready to display details
07-17 18:52:26.781 6243-7329/fr.free.nrw.commons.debug D/MediaDataExtractor: Nominated for deletion: false
07-17 18:52:27.300 6243-7326/fr.free.nrw.commons.debug D/MediaDataExtractor: Nominated for deletion: false
07-17 18:52:27.315 6243-7329/fr.free.nrw.commons.debug D/MediaDetailFragment: java.io.IOException: org.xml.sax.SAXParseException: Unexpected end of document
        at fr.free.nrw.commons.MediaDataExtractor.processWikiParseTree(MediaDataExtractor.java:104)
        at fr.free.nrw.commons.MediaDataExtractor.fetch(MediaDataExtractor.java:77)
        at fr.free.nrw.commons.media.MediaDetailFragment$3.doInBackground(MediaDetailFragment.java:252)
        at fr.free.nrw.commons.media.MediaDetailFragment$3.doInBackground(MediaDetailFragment.java:241)
        at android.os.AsyncTask$2.call(AsyncTask.java:304)
        at java.util.concurrent.FutureTask.run(FutureTask.java:237)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
        at java.lang.Thread.run(Thread.java:761)
     Caused by: org.xml.sax.SAXParseException: Unexpected end of document
        at org.apache.harmony.xml.parsers.DocumentBuilderImpl.parse(DocumentBuilderImpl.java:125)
        at javax.xml.parsers.DocumentBuilder.parse(DocumentBuilder.java:107)
        at fr.free.nrw.commons.MediaDataExtractor.processWikiParseTree(MediaDataExtractor.java:100)
        at fr.free.nrw.commons.MediaDataExtractor.fetch(MediaDataExtractor.java:77) 
        at fr.free.nrw.commons.media.MediaDetailFragment$3.doInBackground(MediaDetailFragment.java:252) 
        at fr.free.nrw.commons.media.MediaDetailFragment$3.doInBackground(MediaDetailFragment.java:241) 
        at android.os.AsyncTask$2.call(AsyncTask.java:304) 
        at java.util.concurrent.FutureTask.run(FutureTask.java:237) 
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607) 
        at java.lang.Thread.run(Thread.java:761) 
07-17 18:52:27.316 6243-6243/fr.free.nrw.commons.debug D/MediaDetailFragment: Failed to load photo details.
07-17 18:52:27.828 6243-7326/fr.free.nrw.commons.debug D/MediaDetailFragment: java.io.IOException: org.xml.sax.SAXParseException: Unexpected end of document
        at fr.free.nrw.commons.MediaDataExtractor.processWikiParseTree(MediaDataExtractor.java:104)
        at fr.free.nrw.commons.MediaDataExtractor.fetch(MediaDataExtractor.java:77)
        at fr.free.nrw.commons.media.MediaDetailFragment$3.doInBackground(MediaDetailFragment.java:252)
        at fr.free.nrw.commons.media.MediaDetailFragment$3.doInBackground(MediaDetailFragment.java:241)
        at android.os.AsyncTask$2.call(AsyncTask.java:304)
        at java.util.concurrent.FutureTask.run(FutureTask.java:237)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
        at java.lang.Thread.run(Thread.java:761)
     Caused by: org.xml.sax.SAXParseException: Unexpected end of document
        at org.apache.harmony.xml.parsers.DocumentBuilderImpl.parse(DocumentBuilderImpl.java:125)
        at javax.xml.parsers.DocumentBuilder.parse(DocumentBuilder.java:107)
        at fr.free.nrw.commons.MediaDataExtractor.processWikiParseTree(MediaDataExtractor.java:100)
        at fr.free.nrw.commons.MediaDataExtractor.fetch(MediaDataExtractor.java:77) 
        at fr.free.nrw.commons.media.MediaDetailFragment$3.doInBackground(MediaDetailFragment.java:252) 
        at fr.free.nrw.commons.media.MediaDetailFragment$3.doInBackground(MediaDetailFragment.java:241) 
        at android.os.AsyncTask$2.call(AsyncTask.java:304) 
        at java.util.concurrent.FutureTask.run(FutureTask.java:237) 
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607) 
        at java.lang.Thread.run(Thread.java:761) 
07-17 18:52:27.829 6243-6243/fr.free.nrw.commons.debug D/MediaDetailFragment: Failed to load photo details.
07-17 18:52:27.835 6243-7329/fr.free.nrw.commons.debug D/MediaDataExtractor: Nominated for deletion: false
07-17 18:52:28.363 6243-7329/fr.free.nrw.commons.debug D/MediaDataExtractor: MediaDataExtractor searching for license

```

--- a/app/src/main/java/fr/free/nrw/commons/auth/AccountUtil.java
+++ b/app/src/main/java/fr/free/nrw/commons/auth/AccountUtil.java
@@ -1,16 +1,53 @@
 package fr.free.nrw.commons.auth;
 
+import android.accounts.Account;
+import android.accounts.AccountManager;
 import android.content.Context;
+import android.support.annotation.Nullable;
+
+import fr.free.nrw.commons.BuildConfig;
+import timber.log.Timber;
 
 public class AccountUtil {
 
-    public static final String ACCOUNT_TYPE = "fr.free.nrw.commons";
     public static final String AUTH_COOKIE = "authCookie";
     public static final String AUTH_TOKEN_TYPE = "CommonsAndroid";
     private final Context context;
 
     public AccountUtil(Context context) {
         this.context = context;
+    }
+
+    /**
+     * @return Account|null
+     */
+    @Nullable
+    public static Account account(Context context) {
+        try {
+            Account[] accounts = accountManager(context).getAccountsByType(BuildConfig.ACCOUNT_TYPE);
+            if (accounts.length > 0) {
+                return accounts[0];
+            }
+        } catch (SecurityException e) {
+            Timber.e(e);
+        }
+        return null;
+    }
+
+    @Nullable
+    public static String getUserName(Context context) {
+        Account account = account(context);
+        return account == null ? null : account.name;
+    }
+
+    @Nullable
+    public static String getPassword(Context context) {
+        Account account = account(context);
+        return account == null ? null : accountManager(context).getPassword(account);
+    }
+
+    private static AccountManager accountManager(Context context) {
+        return AccountManager.get(context);
     }
 
 }

--- a/app/src/main/java/fr/free/nrw/commons/auth/LoginActivity.java
+++ b/app/src/main/java/fr/free/nrw/commons/auth/LoginActivity.java
@@ -55,7 +55,6 @@ import timber.log.Timber;
 import static android.view.KeyEvent.KEYCODE_ENTER;
 import static android.view.View.VISIBLE;
 import static android.view.inputmethod.EditorInfo.IME_ACTION_DONE;
-import static fr.free.nrw.commons.auth.AccountUtil.ACCOUNT_TYPE;
 import static fr.free.nrw.commons.auth.AccountUtil.AUTH_TOKEN_TYPE;
 
 public class LoginActivity extends AccountAuthenticatorActivity {
@@ -242,7 +241,7 @@ public class LoginActivity extends AccountAuthenticatorActivity {
             if (response != null) {
                 Bundle authResult = new Bundle();
                 authResult.putString(AccountManager.KEY_ACCOUNT_NAME, username);
-                authResult.putString(AccountManager.KEY_ACCOUNT_TYPE, ACCOUNT_TYPE);
+                authResult.putString(AccountManager.KEY_ACCOUNT_TYPE, BuildConfig.ACCOUNT_TYPE);
                 response.onResult(authResult);
             }
         }

--- a/app/src/main/java/fr/free/nrw/commons/auth/SessionManager.java
+++ b/app/src/main/java/fr/free/nrw/commons/auth/SessionManager.java
@@ -76,10 +76,6 @@ public class SessionManager {
         ContentResolver.setSyncAutomatically(account, BuildConfig.MODIFICATION_AUTHORITY, true); // Enable sync by default!
     }
 
-    private AccountManager accountManager() {
-        return AccountManager.get(context);
-    }
-
     /**
      * @return Account|null
      */
@@ -95,6 +91,22 @@ public class SessionManager {
         return currentAccount;
     }
 
+    @Nullable
+    public String getUserName() {
+        Account account = getCurrentAccount();
+        return account == null ? null : account.name;
+    }
+
+    @Nullable
+    public String getPassword() {
+        Account account = getCurrentAccount();
+        return account == null ? null : accountManager().getPassword(account);
+    }
+
+    private AccountManager accountManager() {
+        return AccountManager.get(context);
+    }
+
     public Boolean revalidateAuthToken() {
         AccountManager accountManager = AccountManager.get(context);
         Account curAccount = getCurrentAccount();
@@ -103,12 +115,13 @@ public class SessionManager {
             return false; // This should never happen
         }
 
-        accountManager.invalidateAuthToken(BuildConfig.ACCOUNT_TYPE, mediaWikiApi.getAuthCookie());
+        accountManager.invalidateAuthToken(BuildConfig.ACCOUNT_TYPE, null);
         String authCookie = getAuthCookie();
 
         if (authCookie == null) {
             return false;
         }
+
         mediaWikiApi.setAuthCookie(authCookie);
         return true;
     }

--- a/app/src/main/java/fr/free/nrw/commons/auth/WikiAccountAuthenticator.java
+++ b/app/src/main/java/fr/free/nrw/commons/auth/WikiAccountAuthenticator.java
@@ -13,10 +13,7 @@ import android.support.annotation.NonNull;
 import android.support.annotation.Nullable;
 
 import fr.free.nrw.commons.BuildConfig;
-import fr.free.nrw.commons.contributions.ContributionsContentProvider;
-import fr.free.nrw.commons.modifications.ModificationsContentProvider;
 
-import static fr.free.nrw.commons.auth.AccountUtil.ACCOUNT_TYPE;
 import static fr.free.nrw.commons.auth.AccountUtil.AUTH_TOKEN_TYPE;
 
 public class WikiAccountAuthenticator extends AbstractAccountAuthenticator {
@@ -99,7 +96,7 @@ public class WikiAccountAuthenticator extends AbstractAccountAuthenticator {
     }
 
     private boolean supportedAccountType(@Nullable String type) {
-        return ACCOUNT_TYPE.equals(type);
+        return BuildConfig.ACCOUNT_TYPE.equals(type);
     }
 
     private Bundle addAccount(AccountAuthenticatorResponse response) {

--- a/app/src/main/java/fr/free/nrw/commons/mwapi/ApacheHttpClientMediaWikiApi.java
+++ b/app/src/main/java/fr/free/nrw/commons/mwapi/ApacheHttpClientMediaWikiApi.java
@@ -221,7 +221,6 @@ public class ApacheHttpClientMediaWikiApi implements MediaWikiApi {
     @Override
     public String getEditToken() throws IOException {
         String editToken = api.action("query")
-                .param("action", "query")
                 .param("centralauthtoken", getCentralAuthToken())
                 .param("meta", "tokens")
                 .post()

--- a/app/src/main/java/fr/free/nrw/commons/mwapi/ApacheHttpClientMediaWikiApi.java
+++ b/app/src/main/java/fr/free/nrw/commons/mwapi/ApacheHttpClientMediaWikiApi.java
@@ -25,8 +25,6 @@ import org.apache.http.params.BasicHttpParams;
 import org.apache.http.params.CoreProtocolPNames;
 import org.apache.http.util.EntityUtils;
 import org.json.JSONObject;
-import org.mediawiki.api.ApiResult;
-import org.mediawiki.api.MWApi;
 import org.w3c.dom.Element;
 import org.w3c.dom.Node;
 import org.w3c.dom.NodeList;
@@ -37,7 +35,6 @@ import java.net.URL;
 import java.text.ParseException;
 import java.text.SimpleDateFormat;
 import java.util.ArrayList;
-import java.util.Collection;
 import java.util.Collections;
 import java.util.Date;
 import java.util.List;
@@ -48,6 +45,7 @@ import java.util.concurrent.Callable;
 import fr.free.nrw.commons.BuildConfig;
 import fr.free.nrw.commons.Media;
 import fr.free.nrw.commons.PageTitle;
+import fr.free.nrw.commons.auth.AccountUtil;
 import fr.free.nrw.commons.category.CategoryImageUtils;
 import fr.free.nrw.commons.category.QueryContinue;
 import fr.free.nrw.commons.notification.Notification;
@@ -72,8 +70,8 @@ public class ApacheHttpClientMediaWikiApi implements MediaWikiApi {
 
     private static final String THUMB_SIZE = "640";
     private AbstractHttpClient httpClient;
-    private MWApi api;
-    private MWApi wikidataApi;
+    private CustomMwApi api;
+    private CustomMwApi wikidataApi;
     private Context context;
     private SharedPreferences defaultPreferences;
     private SharedPreferences categoryPreferences;
@@ -95,8 +93,8 @@ public class ApacheHttpClientMediaWikiApi implements MediaWikiApi {
         params.setParameter(CoreProtocolPNames.USER_AGENT, getUserAgent());
         httpClient = new DefaultHttpClient(cm, params);
         httpClient.addRequestInterceptor(NetworkInterceptors.getHttpRequestInterceptor());
-        api = new MWApi(apiURL, httpClient);
-        wikidataApi = new MWApi(wikidatApiURL, httpClient);
+        api = new CustomMwApi(apiURL, httpClient);
+        wikidataApi = new CustomMwApi(wikidatApiURL, httpClient);
         this.defaultPreferences = defaultPreferences;
         this.categoryPreferences = categoryPreferences;
         this.gson = gson;
@@ -161,25 +159,25 @@ public class ApacheHttpClientMediaWikiApi implements MediaWikiApi {
     }
 
     /**
-     * @param loginApiResult ApiResult Any clientlogin api result
+     * @param loginCustomApiResult CustomApiResult Any clientlogin api result
      * @return String On success: "PASS"
      * continue: "2FA" (More information required for 2FA)
      * failure: A failure message code (defined by mediawiki)
      * misc:    genericerror-UI, genericerror-REDIRECT, genericerror-RESTART
      */
-    private String getErrorCodeToReturn(ApiResult loginApiResult) {
-        String status = loginApiResult.getString("/api/clientlogin/@status");
+    private String getErrorCodeToReturn(CustomApiResult loginCustomApiResult) {
+        String status = loginCustomApiResult.getString("/api/clientlogin/@status");
         if (status.equals("PASS")) {
             api.isLoggedIn = true;
             setAuthCookieOnLogin(true);
             return status;
         } else if (status.equals("FAIL")) {
             setAuthCookieOnLogin(false);
-            return loginApiResult.getString("/api/clientlogin/@messagecode");
+            return loginCustomApiResult.getString("/api/clientlogin/@messagecode");
         } else if (
                 status.equals("UI")
-                        && loginApiResult.getString("/api/clientlogin/requests/_v/@id").equals("TOTPAuthenticationRequest")
-                        && loginApiResult.getString("/api/clientlogin/requests/_v/@provider").equals("Two-factor authentication (OATH).")
+                        && loginCustomApiResult.getString("/api/clientlogin/requests/_v/@id").equals("TOTPAuthenticationRequest")
+                        && loginCustomApiResult.getString("/api/clientlogin/requests/_v/@provider").equals("Two-factor authentication (OATH).")
                 ) {
             setAuthCookieOnLogin(false);
             return "2FA";
@@ -209,16 +207,27 @@ public class ApacheHttpClientMediaWikiApi implements MediaWikiApi {
     @Override
     public void setAuthCookie(String authCookie) {
         api.setAuthCookie(authCookie);
+
+        Timber.d("Mediawiki auth cookie is %s", api.getAuthCookie());
     }
 
     @Override
     public boolean validateLogin() throws IOException {
-        return api.validateLogin();
+        boolean validateLoginResp = api.validateLogin();
+        Timber.d("Validate login response is %s", validateLoginResp);
+        return validateLoginResp;
     }
 
     @Override
     public String getEditToken() throws IOException {
-        return api.getEditToken();
+        String editToken = api.action("query")
+                .param("action", "query")
+                .param("centralauthtoken", getCentralAuthToken())
+                .param("meta", "tokens")
+                .post()
+                .getString("/api/query/tokens/@csrftoken");
+        Timber.d("MediaWiki edit token is %s", editToken);
+        return editToken;
     }
 
     @Override
@@ -227,6 +236,14 @@ public class ApacheHttpClientMediaWikiApi implements MediaWikiApi {
                 .get()
                 .getString("/api/centralauthtoken/@centralauthtoken");
         Timber.d("MediaWiki Central auth token is %s", centralAuthToken);
+
+        if(centralAuthToken == null || centralAuthToken.isEmpty()) {
+            api.removeAllCookies();
+            String login = login(AccountUtil.getUserName(context), AccountUtil.getPassword(context));
+            if(login.equals("PASS")) {
+                return getCentralAuthToken();
+            }
+        }
         return centralAuthToken;
     }
 
@@ -299,7 +316,7 @@ public class ApacheHttpClientMediaWikiApi implements MediaWikiApi {
     @Override
     @NonNull
     public MediaResult fetchMediaByFilename(String filename) throws IOException {
-        ApiResult apiResult = api.action("query")
+        CustomApiResult apiResult = api.action("query")
                 .param("prop", "revisions")
                 .param("titles", filename)
                 .param("rvprop", "content")
@@ -316,7 +333,7 @@ public class ApacheHttpClientMediaWikiApi implements MediaWikiApi {
     @NonNull
     public Observable<String> searchCategories(String filterValue, int searchCatsLimit) {
         return Single.fromCallable(() -> {
-            List<ApiResult> categoryNodes = null;
+            List<CustomApiResult> categoryNodes = null;
             try {
                 categoryNodes = api.action("query")
                         .param("format", "xml")
@@ -336,7 +353,7 @@ public class ApacheHttpClientMediaWikiApi implements MediaWikiApi {
             }
 
             List<String> categories = new ArrayList<>();
-            for (ApiResult categoryNode : categoryNodes) {
+            for (CustomApiResult categoryNode : categoryNodes) {
                 String cat = categoryNode.getDocument().getTextContent();
                 String catString = cat.replace("Category:", "");
                 categories.add(catString);
@@ -350,7 +367,7 @@ public class ApacheHttpClientMediaWikiApi implements MediaWikiApi {
     @NonNull
     public Observable<String> allCategories(String filterValue, int searchCatsLimit) {
         return Single.fromCallable(() -> {
-            ArrayList<ApiResult> categoryNodes = null;
+            ArrayList<CustomApiResult> categoryNodes = null;
             try {
                 categoryNodes = api.action("query")
                         .param("list", "allcategories")
@@ -367,22 +384,12 @@ public class ApacheHttpClientMediaWikiApi implements MediaWikiApi {
             }
 
             List<String> categories = new ArrayList<>();
-            for (ApiResult categoryNode : categoryNodes) {
+            for (CustomApiResult categoryNode : categoryNodes) {
                 categories.add(categoryNode.getDocument().getTextContent());
             }
 
             return categories;
         }).flatMapObservable(Observable::fromIterable);
-    }
-
-    /**
-     * Get the edit token for making wiki data edits
-     * https://www.mediawiki.org/wiki/API:Tokens
-     * @return
-     * @throws IOException
-     */
-    private String getWikidataEditToken() throws IOException {
-        return wikidataApi.getEditToken();
     }
 
     @Override
@@ -411,7 +418,7 @@ public class ApacheHttpClientMediaWikiApi implements MediaWikiApi {
     @Override
     public String wikidatCreateClaim(String entityId, String property, String snaktype, String value) throws IOException {
         Timber.d("Filename is %s", value);
-        ApiResult result = wikidataApi.action("wbcreateclaim")
+        CustomApiResult result = wikidataApi.action("wbcreateclaim")
                 .param("entity", entityId)
                 .param("centralauthtoken", getCentralAuthToken())
                 .param("token", getWikidataCsrfToken())
@@ -444,7 +451,7 @@ public class ApacheHttpClientMediaWikiApi implements MediaWikiApi {
     @Nullable
     @Override
     public boolean addWikidataEditTag(String revisionId) throws IOException {
-        ApiResult result = wikidataApi.action("tag")
+        CustomApiResult result = wikidataApi.action("tag")
                 .param("revid", revisionId)
                 .param("centralauthtoken", getCentralAuthToken())
                 .param("token", getWikidataCsrfToken())
@@ -471,7 +478,7 @@ public class ApacheHttpClientMediaWikiApi implements MediaWikiApi {
     @NonNull
     public Observable<String> searchTitles(String title, int searchCatsLimit) {
         return Single.fromCallable((Callable<List<String>>) () -> {
-            ArrayList<ApiResult> categoryNodes;
+            ArrayList<CustomApiResult> categoryNodes;
 
             try {
                 categoryNodes = api.action("query")
@@ -493,7 +500,7 @@ public class ApacheHttpClientMediaWikiApi implements MediaWikiApi {
             }
 
             List<String> titleCategories = new ArrayList<>();
-            for (ApiResult categoryNode : categoryNodes) {
+            for (CustomApiResult categoryNode : categoryNodes) {
                 String cat = categoryNode.getDocument().getTextContent();
                 String catString = cat.replace("Category:", "");
                 titleCategories.add(catString);
@@ -506,7 +513,7 @@ public class ApacheHttpClientMediaWikiApi implements MediaWikiApi {
     @Override
     @NonNull
     public LogEventResult logEvents(String user, String lastModified, String queryContinue, int limit) throws IOException {
-        org.mediawiki.api.MWApi.RequestBuilder builder = api.action("query")
+        CustomMwApi.RequestBuilder builder = api.action("query")
                 .param("list", "logevents")
                 .param("letype", "upload")
                 .param("leprop", "title|timestamp|ids")
@@ -518,7 +525,7 @@ public class ApacheHttpClientMediaWikiApi implements MediaWikiApi {
         if (!TextUtils.isEmpty(queryContinue)) {
             builder.param("lestart", queryContinue);
         }
-        ApiResult result = builder.get();
+        CustomApiResult result = builder.get();
 
         return new LogEventResult(
                 getLogEventsFromResult(result),
@@ -526,11 +533,11 @@ public class ApacheHttpClientMediaWikiApi implements MediaWikiApi {
     }
 
     @NonNull
-    private ArrayList<LogEventResult.LogEvent> getLogEventsFromResult(ApiResult result) {
-        ArrayList<ApiResult> uploads = result.getNodes("/api/query/logevents/item");
+    private ArrayList<LogEventResult.LogEvent> getLogEventsFromResult(CustomApiResult result) {
+        ArrayList<CustomApiResult> uploads = result.getNodes("/api/query/logevents/item");
         Timber.d("%d results!", uploads.size());
         ArrayList<LogEventResult.LogEvent> logEvents = new ArrayList<>();
-        for (ApiResult image : uploads) {
+        for (CustomApiResult image : uploads) {
             logEvents.add(new LogEventResult.LogEvent(
                     image.getString("@pageid"),
                     image.getString("@title"),
@@ -554,7 +561,7 @@ public class ApacheHttpClientMediaWikiApi implements MediaWikiApi {
     @Override
     @NonNull
     public List<Notification> getNotifications() {
-        ApiResult notificationNode = null;
+        CustomApiResult notificationNode = null;
         try {
             notificationNode = api.action("query")
                     .param("notprop", "list")
@@ -589,9 +596,9 @@ public class ApacheHttpClientMediaWikiApi implements MediaWikiApi {
     @Override
     @NonNull
     public List<String> getSubCategoryList(String categoryName) {
-        ApiResult apiResult = null;
+        CustomApiResult apiResult = null;
         try {
-            MWApi.RequestBuilder requestBuilder = api.action("query")
+            CustomMwApi.RequestBuilder requestBuilder = api.action("query")
                     .param("generator", "categorymembers")
                     .param("format", "xml")
                     .param("gcmtype","subcat")
@@ -609,7 +616,7 @@ public class ApacheHttpClientMediaWikiApi implements MediaWikiApi {
             return new ArrayList<>();
         }
 
-        ApiResult categoryImagesNode = apiResult.getNode("/api/query/pages");
+        CustomApiResult categoryImagesNode = apiResult.getNode("/api/query/pages");
         if (categoryImagesNode == null
                 || categoryImagesNode.getDocument() == null
                 || categoryImagesNode.getDocument().getChildNodes() == null
@@ -630,9 +637,9 @@ public class ApacheHttpClientMediaWikiApi implements MediaWikiApi {
     @Override
     @NonNull
     public List<String> getParentCategoryList(String categoryName) {
-        ApiResult apiResult = null;
+        CustomApiResult apiResult = null;
         try {
-            MWApi.RequestBuilder requestBuilder = api.action("query")
+            CustomMwApi.RequestBuilder requestBuilder = api.action("query")
                     .param("generator", "categories")
                     .param("format", "xml")
                     .param("titles", categoryName)
@@ -649,7 +656,7 @@ public class ApacheHttpClientMediaWikiApi implements MediaWikiApi {
             return new ArrayList<>();
         }
 
-        ApiResult categoryImagesNode = apiResult.getNode("/api/query/pages");
+        CustomApiResult categoryImagesNode = apiResult.getNode("/api/query/pages");
         if (categoryImagesNode == null
                 || categoryImagesNode.getDocument() == null
                 || categoryImagesNode.getDocument().getChildNodes() == null
@@ -672,9 +679,9 @@ public class ApacheHttpClientMediaWikiApi implements MediaWikiApi {
     @Override
     @NonNull
     public List<Media> getCategoryImages(String categoryName) {
-        ApiResult apiResult = null;
+        CustomApiResult apiResult = null;
         try {
-            MWApi.RequestBuilder requestBuilder = api.action("query")
+            CustomMwApi.RequestBuilder requestBuilder = api.action("query")
                     .param("generator", "categorymembers")
                     .param("format", "xml")
                     .param("gcmtype", "file")
@@ -699,7 +706,7 @@ public class ApacheHttpClientMediaWikiApi implements MediaWikiApi {
             return new ArrayList<>();
         }
 
-        ApiResult categoryImagesNode = apiResult.getNode("/api/query/pages");
+        CustomApiResult categoryImagesNode = apiResult.getNode("/api/query/pages");
         if (categoryImagesNode == null
                 || categoryImagesNode.getDocument() == null
                 || categoryImagesNode.getDocument().getChildNodes() == null
@@ -727,7 +734,7 @@ public class ApacheHttpClientMediaWikiApi implements MediaWikiApi {
     @Override
     @NonNull
     public List<Media> searchImages(String query, int offset) {
-        List<ApiResult> imageNodes = null;
+        List<CustomApiResult> imageNodes = null;
         try {
             imageNodes = api.action("query")
                     .param("format", "xml")
@@ -748,7 +755,7 @@ public class ApacheHttpClientMediaWikiApi implements MediaWikiApi {
         }
 
         List<Media> images = new ArrayList<>();
-        for (ApiResult imageNode : imageNodes) {
+        for (CustomApiResult imageNode : imageNodes) {
             String imgName = imageNode.getDocument().getTextContent();
             images.add(new Media(imgName));
         }
@@ -765,7 +772,7 @@ public class ApacheHttpClientMediaWikiApi implements MediaWikiApi {
     @Override
     @NonNull
     public List<String> searchCategory(String query, int offset) {
-        List<ApiResult> categoryNodes = null;
+        List<CustomApiResult> categoryNodes = null;
         try {
             categoryNodes = api.action("query")
                     .param("format", "xml")
@@ -786,7 +793,7 @@ public class ApacheHttpClientMediaWikiApi implements MediaWikiApi {
         }
 
         List<String> categories = new ArrayList<>();
-        for (ApiResult categoryNode : categoryNodes) {
+        for (CustomApiResult categoryNode : categoryNodes) {
             String catName = categoryNode.getDocument().getTextContent();
             categories.add(catName);
         }
@@ -858,11 +865,11 @@ public class ApacheHttpClientMediaWikiApi implements MediaWikiApi {
                                    long dataLength,
                                    String pageContents,
                                    String editSummary,
-                                   final ProgressListener progressListener,
                                    Uri fileUri,
-                                   Uri contentProviderUri) throws IOException {
+                                   Uri contentProviderUri,
+                                   final ProgressListener progressListener) throws IOException {
 
-        ApiResult result = api.upload(filename, file, dataLength, pageContents, editSummary, progressListener::onProgress);
+        CustomApiResult result = api.upload(filename, file, dataLength, pageContents, editSummary, getCentralAuthToken(), getEditToken(), progressListener::onProgress);
 
         Log.e("WTF", "Result: " + result.toString());
 
@@ -910,7 +917,7 @@ public class ApacheHttpClientMediaWikiApi implements MediaWikiApi {
     public boolean isUserBlockedFromCommons() {
         boolean userBlocked = false;
         try {
-            ApiResult result = api.action("query")
+            CustomApiResult result = api.action("query")
                     .param("action", "query")
                     .param("format", "xml")
                     .param("meta", "userinfo")

--- a/app/src/main/java/fr/free/nrw/commons/mwapi/CustomApiResult.java
+++ b/app/src/main/java/fr/free/nrw/commons/mwapi/CustomApiResult.java
@@ -1,0 +1,123 @@
+package fr.free.nrw.commons.mwapi;
+
+import org.apache.http.client.HttpClient;
+import org.w3c.dom.Document;
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
+import org.xml.sax.SAXException;
+
+import java.io.IOError;
+import java.io.IOException;
+import java.io.StringWriter;
+import java.util.ArrayList;
+
+import javax.xml.parsers.DocumentBuilder;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.ParserConfigurationException;
+import javax.xml.transform.Transformer;
+import javax.xml.transform.TransformerException;
+import javax.xml.transform.TransformerFactory;
+import javax.xml.transform.dom.DOMSource;
+import javax.xml.transform.stream.StreamResult;
+import javax.xml.xpath.XPath;
+import javax.xml.xpath.XPathConstants;
+import javax.xml.xpath.XPathExpressionException;
+import javax.xml.xpath.XPathFactory;
+
+import in.yuvi.http.fluent.Http;
+import timber.log.Timber;
+
+public class CustomApiResult {
+    private Node doc;
+    private XPath evaluator;
+
+    CustomApiResult(Node doc) {
+        this.doc = doc;
+        this.evaluator = XPathFactory.newInstance().newXPath();
+    }
+
+    static CustomApiResult fromRequestBuilder(Http.HttpRequestBuilder builder, HttpClient client) throws IOException {
+
+        Timber.d("API request is %s", builder.toString());
+        Timber.d("API params are %s", client.getParams());
+
+        try {
+            DocumentBuilder docBuilder = DocumentBuilderFactory.newInstance().newDocumentBuilder();
+            Document doc = docBuilder.parse(builder.use(client).charset("utf-8").data("format", "xml").asResponse().getEntity().getContent());
+            printStringFromDocument(doc);
+            return new CustomApiResult(doc);
+        } catch (ParserConfigurationException e) {
+            // I don't know wtf I can do about this on...
+            throw new RuntimeException(e);
+        } catch (IllegalStateException e) {
+            // So, this should never actually happen - since we assume MediaWiki always generates valid json
+            // So the only thing causing this would be a network truncation
+            // Sooo... I can throw IOError
+            // Thanks Java, for making me spend significant time on shit that happens once in a bluemoon
+            // I surely am writing Nuclear Submarine controller code
+            throw new IOError(e);
+        } catch (SAXException e) {
+            // See Rant above
+            throw new IOError(e);
+        }
+    }
+
+    public static void printStringFromDocument(Document doc)
+    {
+        try
+        {
+            DOMSource domSource = new DOMSource(doc);
+            StringWriter writer = new StringWriter();
+            StreamResult result = new StreamResult(writer);
+            TransformerFactory tf = TransformerFactory.newInstance();
+            Transformer transformer = tf.newTransformer();
+            transformer.transform(domSource, result);
+            Timber.d("API response is\n %s", writer.toString());
+        }
+        catch(TransformerException ex)
+        {
+            Timber.d("Error occurred in transforming", ex);
+        }
+    }
+
+    public Node getDocument() {
+        return doc;
+    }
+
+    public ArrayList<CustomApiResult> getNodes(String xpath) {
+        try {
+            ArrayList<CustomApiResult> results = new ArrayList<CustomApiResult>();
+            NodeList nodes = (NodeList) evaluator.evaluate(xpath, doc, XPathConstants.NODESET);
+            for(int i = 0; i < nodes.getLength(); i++) {
+                results.add(new CustomApiResult(nodes.item(i)));
+            }
+            return results;
+        } catch (XPathExpressionException e) {
+            return null;
+        }
+
+    }
+    public CustomApiResult getNode(String xpath) {
+        try {
+            return new CustomApiResult((Node) evaluator.evaluate(xpath, doc, XPathConstants.NODE));
+        } catch (XPathExpressionException e) {
+            return null;
+        }
+    }
+
+    public Double getNumber(String xpath) {
+        try {
+            return (Double) evaluator.evaluate(xpath, doc, XPathConstants.NUMBER);
+        } catch (XPathExpressionException e) {
+            return null;
+        }
+    }
+
+    public String getString(String xpath) {
+        try {
+            return (String) evaluator.evaluate(xpath, doc, XPathConstants.STRING);
+        } catch (XPathExpressionException e) {
+            return null;
+        }
+    }
+}

--- a/app/src/main/java/fr/free/nrw/commons/mwapi/CustomMwApi.java
+++ b/app/src/main/java/fr/free/nrw/commons/mwapi/CustomMwApi.java
@@ -1,0 +1,183 @@
+package fr.free.nrw.commons.mwapi;
+
+import org.apache.http.cookie.Cookie;
+import org.apache.http.impl.client.AbstractHttpClient;
+import org.apache.http.impl.cookie.BasicClientCookie;
+import org.mediawiki.api.ApiResult;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.HashMap;
+import java.util.List;
+
+import in.yuvi.http.fluent.Http;
+import in.yuvi.http.fluent.ProgressListener;
+import timber.log.Timber;
+
+public class CustomMwApi {
+    public class RequestBuilder {
+        private HashMap<String, Object> params;
+        private CustomMwApi api;
+
+        RequestBuilder(CustomMwApi api) {
+            params = new HashMap<String, Object>();
+            this.api = api;
+        }
+
+        public RequestBuilder param(String key, Object value) {
+            params.put(key, value);
+            return this;
+        }
+
+        public CustomApiResult get() throws IOException {
+            return api.makeRequest("GET", params);
+        }
+
+        public CustomApiResult post() throws IOException {
+            return api.makeRequest("POST", params);
+        }
+    }
+
+    private AbstractHttpClient client;
+    private String apiURL;
+    public boolean isLoggedIn;
+    private String authCookie = null;
+    private String userName = null;
+    private String userID = null;
+
+    public CustomMwApi(String apiURL, AbstractHttpClient client) {
+        this.apiURL = apiURL;
+        this.client = client;
+    }
+
+    public RequestBuilder action(String action) {
+        RequestBuilder builder = new RequestBuilder(this);
+        builder.param("action", action);
+        return builder;
+    }
+
+    public String getAuthCookie() {
+        if(authCookie == null){
+            authCookie = "";
+            List<Cookie> cookies = client.getCookieStore().getCookies();
+            for(Cookie cookie: cookies) {
+                authCookie += cookie.getName() + "=" + cookie.getValue() + ";";
+            }
+        }
+        return authCookie;
+    }
+
+    public void setAuthCookie(String authCookie) {
+        this.authCookie = authCookie;
+        this.isLoggedIn = true;
+        String[] cookies = authCookie.split(";");
+        String domain;
+        try {
+            domain = new URL(apiURL).getHost();
+        } catch (MalformedURLException e) {
+            // Mighty well better not happen!
+            e.printStackTrace();
+            throw new RuntimeException(e);
+        }
+        // This works because I know which cookies are going to be set by MediaWiki, and they don't contain a = or ; in them :D
+        for(String cookie: cookies) {
+            String[] parts = cookie.split("=");
+            BasicClientCookie c = new BasicClientCookie(parts[0], parts[1]);
+            c.setDomain(domain);
+            client.getCookieStore().addCookie(c);
+        }
+    }
+
+    public void removeAllCookies() {
+        client.getCookieStore().clear();
+    }
+
+    public boolean validateLogin() throws IOException {
+        CustomApiResult userMeta = this.action("query").param("meta", "userinfo").get();
+        this.userID = userMeta.getString("/api/query/userinfo/@id");
+        this.userName = userMeta.getString("/api/query/userinfo/@name");
+        Timber.d("User id is %s and user name is %s", userID, userName);
+        return !userID.equals("0");
+    }
+
+    public String getUserID() throws IOException {
+        if(this.userID == null || this.userID.equals("0")) {
+            this.validateLogin();
+        }
+        return userID;
+    }
+
+    public String getUserName() throws IOException {
+        if(this.userID == null || this.userID.equals("0")) {
+            this.validateLogin();
+        }
+        return userName;
+    }
+
+    public String login(String username, String password) throws IOException {
+        CustomApiResult tokenData = this.action("login").param("lgname", username).param("lgpassword", password).post();
+        String result = tokenData.getString("/api/login/@result");
+        if (result.equals("NeedToken")) {
+            String token = tokenData.getString("/api/login/@token");
+            CustomApiResult confirmData = this.action("login").param("lgname", username).param("lgpassword", password).param("lgtoken", token).post();
+            String finalResult = confirmData.getString("/api/login/@result");
+            if(finalResult.equals("Success")) {
+                isLoggedIn = true;
+            }
+            return finalResult;
+        } else {
+            return result;
+        }
+    }
+
+    public CustomApiResult upload(String filename, InputStream file, long length, String text, String comment, String centralAuthToken, String token) throws IOException {
+        return this.upload(filename, file, length, text, comment,centralAuthToken, token, null);
+    }
+
+    public CustomApiResult upload(String filename, InputStream file, String text, String comment, String centralAuthToken, String token) throws IOException {
+        return this.upload(filename, file, -1, text, comment,centralAuthToken, token, null);
+    }
+
+    public CustomApiResult upload(String filename, InputStream file, long length, String text, String comment, String centralAuthToken, String token, ProgressListener uploadProgressListener) throws IOException {
+        Timber.d("Token being used is %s", token);
+
+        Http.HttpRequestBuilder builder = Http.multipart(apiURL)
+                .data("action", "upload")
+                .data("token", token)
+                .data("centralauthtoken", centralAuthToken)
+                .data("text", text)
+                .data("ignorewarnings", "1")
+                .data("comment", comment)
+                .data("filename", filename)
+                .sendProgressListener(uploadProgressListener);
+        if(length != -1) {
+            builder.file("file", filename, file, length);
+        } else {
+            builder.file("file", filename, file);
+        }
+
+        Timber.d("Final cookies are %s", client.getCookieStore().getCookies().toString());
+
+        return CustomApiResult.fromRequestBuilder(builder, client);
+    }
+
+    public void logout() throws IOException {
+        // I should be doing more validation here, but meh
+        isLoggedIn = false;
+        this.action("logout").post();
+    }
+
+    private CustomApiResult makeRequest(String method, HashMap<String, Object> params) throws IOException {
+        Http.HttpRequestBuilder builder;
+        if (method.equals("POST")) {
+            builder = Http.post(apiURL);
+        } else {
+            builder = Http.get(apiURL);
+        }
+        builder.data(params);
+        return CustomApiResult.fromRequestBuilder(builder, client);
+    }
+}
+;

--- a/app/src/main/java/fr/free/nrw/commons/mwapi/MediaWikiApi.java
+++ b/app/src/main/java/fr/free/nrw/commons/mwapi/MediaWikiApi.java
@@ -55,7 +55,7 @@ public interface MediaWikiApi {
     List<String> searchCategory(String title, int offset);
 
     @NonNull
-    UploadResult uploadFile(String filename, InputStream file, long dataLength, String pageContents, String editSummary, ProgressListener progressListener, Uri fileUri, Uri contentProviderUri) throws IOException;
+    UploadResult uploadFile(String filename, InputStream file, long dataLength, String pageContents, String editSummary, Uri fileUri, Uri contentProviderUri, ProgressListener progressListener) throws IOException;
 
     @Nullable
     String edit(String editToken, String processedPageContent, String filename, String summary) throws IOException;

--- a/app/src/main/java/fr/free/nrw/commons/theme/NavigationBaseActivity.java
+++ b/app/src/main/java/fr/free/nrw/commons/theme/NavigationBaseActivity.java
@@ -71,7 +71,7 @@ public abstract class NavigationBaseActivity extends BaseActivity
         View navHeaderView = navigationView.getHeaderView(0);
         TextView username = navHeaderView.findViewById(R.id.username);
         AccountManager accountManager = AccountManager.get(this);
-        Account[] allAccounts = accountManager.getAccountsByType(AccountUtil.ACCOUNT_TYPE);
+        Account[] allAccounts = accountManager.getAccountsByType(BuildConfig.ACCOUNT_TYPE);
         if (allAccounts.length != 0) {
             username.setText(allAccounts[0].name);
         }

--- a/app/src/main/java/fr/free/nrw/commons/upload/UploadService.java
+++ b/app/src/main/java/fr/free/nrw/commons/upload/UploadService.java
@@ -249,7 +249,7 @@ public class UploadService extends HandlerService<Contribution> {
                     getString(R.string.upload_progress_notification_title_finishing, contribution.getDisplayTitle()),
                     contribution
             );
-            UploadResult uploadResult = mwApi.uploadFile(filename, fileInputStream, contribution.getDataLength(), contribution.getPageContents(), contribution.getEditSummary(), notificationUpdater, contribution.getLocalUri(), contribution.getContentProviderUri());
+            UploadResult uploadResult = mwApi.uploadFile(filename, fileInputStream, contribution.getDataLength(), contribution.getPageContents(), contribution.getEditSummary(), contribution.getLocalUri(), contribution.getContentProviderUri(), notificationUpdater);
 
             Timber.d("Response is %s", uploadResult.toString());
 

--- a/app/src/test/kotlin/fr/free/nrw/commons/mwapi/ApacheHttpClientMediaWikiApiTest.kt
+++ b/app/src/test/kotlin/fr/free/nrw/commons/mwapi/ApacheHttpClientMediaWikiApiTest.kt
@@ -182,15 +182,23 @@ class ApacheHttpClientMediaWikiApiTest {
 
     @Test
     fun editToken() {
-        server.enqueue(MockResponse().setBody("<?xml version=\"1.0\"?><api><tokens edittoken=\"baz\" /></api>"))
+        server.enqueue(MockResponse().setBody("<?xml version=\"1.0\"?><api><centralauthtoken centralauthtoken=\"abc\" /></api>"))
+        server.enqueue(MockResponse().setBody("<?xml version=\"1.0\"?><api><query><tokens csrftoken=\"baz\" /></query></api>"))
 
         val result = testObject.editToken
 
-        assertBasicRequestParameters(server, "GET").let { loginTokenRequest ->
-            parseQueryParams(loginTokenRequest).let { params ->
+        assertBasicRequestParameters(server, "GET").let { centralAuthTokenRequest ->
+            parseQueryParams(centralAuthTokenRequest).let { params ->
                 assertEquals("xml", params["format"])
-                assertEquals("tokens", params["action"])
-                assertEquals("edit", params["type"])
+                assertEquals("centralauthtoken", params["action"])
+            }
+        }
+
+        assertBasicRequestParameters(server, "POST").let { editTokenRequest ->
+            parseBody(editTokenRequest.body.readUtf8()).let { body ->
+                assertEquals("query", body["action"])
+                assertEquals("abc", body["centralauthtoken"])
+                assertEquals("tokens", body["meta"])
             }
         }
 

